crates/language/src/language.rs

```diff
@@ -890,20 +892,28 @@ impl LanguageRegistry {
         })
     }
 
+    /// context_info is a (worktree ID, &AppContext),
+    /// which will be used to load user's custom file association settings.
     pub fn language_for_file(
         self: &Arc<Self>,
         path: impl AsRef<Path>,
+        context_info: Option<(usize, &AppContext)>,
         content: Option<&Rope>,
     ) -> UnwrapFuture<oneshot::Receiver<Result<Arc<Language>>>> {
         let path = path.as_ref();
         let filename = path.file_name().and_then(|name| name.to_str());
         let extension = path.extension_or_hidden_file_name();
         let path_suffixes = [extension, filename];
+        let associations = context_info.map(|(worktree_id, cx)| {
+            &AllLanguageSettings::get(Some((worktree_id, path)), cx).associations
+        });
         self.get_or_load_language(|config| {
-            let path_matches = config
-                .path_suffixes
-                .iter()
-                .any(|suffix| path_suffixes.contains(&Some(suffix.as_str())));
+            // Give higher priority to user's custom settings
+            let path_matches = associations
+                .and_then(|associations| associations.get(&config.name))
```

Comment:
One thing that really bugs me is the ability to spell languages differently in the config and zero feedback Zed provides for that now: so if somebody writes `rust` or `typescript`, there will be zero warnings and just a silent noop instead of applying the settings.

I think we should check the entire config and `log::error` any missing languages at least.
Good level of support would be to provide additions to the corresponding `JsonSchema` so that the user could autocomplete language names.